Configuration: Add option to automatically load a config seed from a buildconf

planthaber opened the pull request:
Allows automatic loading a default "seed-config.yml" file from the buildconf repository
```ruby
Autoproj.config.load_config_once('seed-config.yml')
```
in a buildconf init.rb to apply a configuration seed from the same buildconf repository

The second call will ask the user if the default should be applied or not:
```ruby
Autoproj.config.load_config_once_with_permission('seed-config.yml')
```
This will load the autoproj/seed-config.yml while bootstrapping (to avoid questions asked)

Discussed here: https://github.com/rock-core/autoproj/issues/364

doudou reviewed (commented):
Since this is an instance method, it should rather obviously act on itself instead of Autoproj.config. And `configuration_option` as well as `user_config` are now instance methods on the same class (named declare and get I think)

Instead of referring to the global Autoproj.workspace, let the caller provide you with a `root_path` keyword argument whose default is `Autoproj.workspace.root`

doudou commented on lib/autoproj/configuration.rb:
```diff
@@ -606,5 +606,34 @@ def to_hash
+        # The filename parameter is the name of the config seed yml file in the repository
+        def load_config_once(filename)
+            unless Autoproj.config.has_value_for?("default_config_applied_#{filename}")
+                seed_config = File.join(Autoproj.workspace.root_dir, "autoproj", filename)
```
Do not create massive assumptions on where `filename` is. Using `File.expand_path(filename, root_dir)` would allow the caller to provide a full path. Where root_dir's default is `Autoproj.workspace.config_dir`

doudou commented on lib/autoproj/configuration.rb:
```diff
@@ -606,5 +606,34 @@ def to_hash
+        # config option, e.g. the git server configuration settings
+        # The filename parameter is the name of the config seed yml file in the repository
+        def load_config_once(filename)
+            unless Autoproj.config.has_value_for?("default_config_applied_#{filename}")
```
Do an early return instead of a gigantic if.

~~~
return if get("default_config_applied_#{filename}", false)

<rest of method>
~~~

doudou commented:
Please write unit tests for this ...

planthaber commented:
Updated the PR (added tests and added the suggested stuff)

I realized that due to the "load once" rule the user is asked if the config should be applied when calling "autoproj reconfigure" but it is not applied because the config said is was already applied.

I think this is actually "good" behavior, otherwise the default could overwrite question asked before the defaults are applied, so I now prevent asking for that option when calling reconfigure

doudou commented on lib/autoproj/configuration.rb:
```diff
@@ -606,5 +606,35 @@ def to_hash
+        # The call to this function has to be in the init.rb of the buildconf BEFORE any other
+        # config option, e.g. the git server configuration settings
+        # The filename parameter is the name of the config seed yml file in the repository
+        def load_config_once(filename)
+            return if get("default_config_applied_#{filename}", false)
+
+            seed_config = File.expand_path(filename, Autoproj.workspace.config_dir)
```
What I meant in my earlier review is that the root_dir should be passed as a keyword argument, that is

```
def load_config_once(filename, root_dir: Autoproj.workspace.config_dir)
```

doudou commented on test/test_configuration.rb:
```diff
@@ -433,5 +433,61 @@ module Autoproj
         end
+
+        describe "#load_config_once" do
+            it "load config once loads config only once" do
```
Please add a test for the root_dir argument (i.e. resolution of relative paths)

planthaber commented:
Updated the PR

doudou commented:
In this case, I would expect to be the one merging as I and @g-arjones are the de-facto maintainer of this package. If you had a lot of contributions, I would expect you to merge.

doudou commented:
All good. Thanks !
